python/src/pyarrow/adapters/pandas.cc

```diff
@@ -548,6 +548,31 @@ class ArrowDeserializer {
     return Status::OK();
   }
 
+  Status OutputFromData(int type, void* data) {
+    // Zero-Copy. We can pass the data pointer directly to NumPy.
+    Py_INCREF(py_ref_);
+    npy_intp dims[1] = {col_->length()};
+    out_ = reinterpret_cast<PyArrayObject*>(PyArray_SimpleNewFromData(1, dims,
+                type, data));
+
+    if (out_ == NULL) {
+      // Error occurred, trust that SimpleNew set the error state
+      Py_DECREF(py_ref_);
+      return Status::OK();
+    }
+
+    if (PyArray_SetBaseObject(out_, py_ref_) == -1) {
+      // Error occurred, trust that SetBaseObject set the error state
+      Py_DECREF(py_ref_);
+      return Status::OK();
+    }
+
+    // Arrow data is immutable.
+    PyArray_CLEARFLAGS(out_, NPY_ARRAY_WRITEABLE);
```

Comment:
I'm not sure this contract should be enforced here (basically a caveat emptor to the caller of this to-Arrow conversion). But I don't feel extremely strongly about it